[Agents] widget markdown renderer

kalvin807 opened the pull request:
Adopted streamdown and trimming it down to basic feature.

Small things
- added bundle analyzer 
- added new dev page for markdown render `http://localhost:5173/markdown`

Streamdown vs this
```
* ✅ **Strong text**
* ✅ Inline code
* ✅ Links (`<a>`)
* ✅ Autolink literals
* ✅ Superscript
* ✅ Subscript
* ✅ Strikethrough (GFM)
* ✅ H1-H6
* ✅ Ordered lists
* ✅ Unordered lists
* ✅ List items
* ✅ Task lists
* ✅ Table
  * ✅ Table copy button
  * ❌ Table download dropdown (Removed, difficult to use in small widget, need a design if needed)
* ✅ Inline code
* ✅ Code blocks
* ⚠️ Language-specific highlighting
  * only light theme: `github-light`
  * only `typescript, javascript, python, markdown`
* ✅ Code copy button
* ✅ Paragraphs
* ✅ Horizontal rules
* ✅ Blockquotes
* ✅ Images
  * ✅ Download (some images do not work due to CORS, maybe we can remove? user can right click to save anyway)
* ✅ Footnotes
* ❌ Math equation
* ❌ Mermaid diagram
```

### TODO
- designer check ✅ 
- apply to widget (in next PR)

## bundle size 
``` 
1. Original 
	1. Unzipped 627.79 KB
	2. GZip: 181.03 KB
2. With streamdown pkg
	1. Unzipped 13.87 MB
	2. Gzip 3.78 MB
3. With diy streamdown lite
	1. Unzipped 1.48 MB
	2. Gzip 381 KB
```
can be even smaller if we drop syntax highlight

<img width="3456" height="1996" alt="CleanShot 2025-11-12 at 17 08 39@2x" src="https://github.com/user-attachments/assets/37582e00-401b-4d75-91ea-c2fc92a9fa9a" />




https://github.com/user-attachments/assets/320e914b-b759-4e21-a727-40de6fa68d0d

kalvin807 commented on packages/convai-widget-core/src/utils/cn.ts:
```diff
@@ -0,0 +1,4 @@
+import { type ClassValue, clsx } from "clsx";
+import { twMerge } from "tailwind-merge";
+
+export const cn = (...inputs: ClassValue[]) => twMerge(clsx(inputs));
```
📝 `twMerge` make stacking tailwind classname easier

AngeloGiacco commented:
yo this is huge, great stuff also on the bundler !

kamilk-11 commented on packages/convai-widget-core/src/markdown/components/CodeBlock.tsx:
```diff
@@ -0,0 +1,220 @@
+
+    this.initializationPromise = this.ensureHighlighterInitialized();
+    await this.initializationPromise;
+    this.initializationPromise = null;
```
You can simplify a lot of this code by using the fact that promises already serve as a value cache, basically:

```ts
private highlighter: Promise<HighlighterCore> | null = null;

private getHighlighter(): Promise<HighlighterCore> {
  this.highlighter ??= createHighlighter();
  return this.highlighter;
}

async highlightCode(/* ... */): Promise<string> {
  const highlighter = await this.getHighlighter();
  // ...
}
```

kamilk-11 commented on packages/convai-widget-core/src/markdown/components/components.tsx:
```diff
@@ -0,0 +1,720 @@
+      {children}
+    </h3>
+  ),
+  (p, n) => sameClassAndNode(p, n)
```
I'm guessing this is copy-paste from streamdown but could be just:
```suggestion
  sameClassAndNode
```

kamilk-11 commented on packages/convai-widget-core/src/markdown/components/InfoCard.tsx:
```diff
@@ -0,0 +1,92 @@
+}
+
+type InfoCardActionsProps = {
+  actions: ActionConfig[];
```
This should just accept `ComponentChildren` and have the components pass buttons here in a fragment
`ActionConfig` is just reimplementing JSX tbh
```suggestion
  actions: ComponentChildren;
```

kamilk-11 commented on packages/convai-widget-core/src/markdown/components/InfoCard.tsx:
```diff
@@ -0,0 +1,92 @@
+            icon={action.icon}
+            onClick={action.onClick}
+            iconClassName="text-sm"
+            variant="md-button"
```
Can we apply `text-sm` to icons automatically when `variant `is `md-button`?

kamilk-11 commented on packages/convai-widget-core/src/markdown/components/InfoCard.tsx:
```diff
@@ -0,0 +1,92 @@
+          className={actionsClassName}
+        />
+      )}
+      <div className={cn(className)}>{children}</div>
```
```suggestion
      <div className={className}>{children}</div>
```

kamilk-11 reviewed (commented):
Nice! Leaving some suggestions
Unsure what is the plan for dynamic imports though

kalvin807 commented:
@kamilk-11 please take a look again when free

- Rewrote code-block using lezer (1.33mb with more language support) + option to wrap line
<img width="3456" height="1984" alt="CleanShot 2025-11-17 at 11 42 58@2x" src="https://github.com/user-attachments/assets/5a5545c7-8666-43c3-a5b7-bc2297abc0fb" />

📝 I tired highlightjs, they can be even smaller but in some case it is not highlighting correctly, https://github.com/highlightjs/highlight.js/issues/3625#issuecomment-2578161714

- Rewrote InfoCard into ContentBlock with radix-ui like composition
- Applied comments


supported lang now
```
  html: htmlParser,
  js: javascriptParser.configure({ dialect: "js" }),
  jsx: javascriptParser.configure({ dialect: "jsx" }),
  ts: javascriptParser.configure({ dialect: "ts" }),
  tsx: javascriptParser.configure({ dialect: "tsx" }),
  json: jsonParser,
  md: markdownParser,
  py: pythonParser,
  xml: xmlParser,
  yaml: yamlParser,
```

kalvin807 commented on packages/convai-widget-core/src/markdown/utils/highlighter.tsx:
```diff
@@ -0,0 +1,75 @@
+  language?: string;
+}
+
+export const Code = (props: CodeProps) => {
```
fixed
